src/pylibsshext/channel.pyx

```diff
@@ -133,37 +133,31 @@ cdef class Channel:
         return response
 
     def exec_command(self, command):
-        cdef libssh.ssh_channel channel = libssh.ssh_channel_new(self._libssh_session)
-        if channel is NULL:
-            raise MemoryError
-        rc = libssh.ssh_channel_open_session(channel)
-        if rc != libssh.SSH_OK:
-            libssh.ssh_channel_free(channel)
-            raise CalledProcessError()
+        rc = libssh.ssh_channel_request_exec(self._libssh_channel, command.encode("utf-8"))
 
-        rc = libssh.ssh_channel_request_exec(channel, command.encode("utf-8"))
         if rc != libssh.SSH_OK:
-            libssh.ssh_channel_free(channel)
+            self.close()
             raise CalledProcessError()
+        result = CompletedProcess(args=command, returncode=-1, stdout=b'', stderr=b'')
 
         cdef callbacks.ssh_channel_callbacks_struct cb
         memset(&cb, 0, sizeof(cb))
         cb.channel_data_function = <callbacks.ssh_channel_data_callback>&_process_outputs
-        result = CompletedProcess(args=command, returncode=-1, stdout=b'', stderr=b'')
         cb.userdata = <void *>result
         callbacks.ssh_callbacks_init(&cb)
-        callbacks.ssh_set_channel_callbacks(channel, &cb)
+        callbacks.ssh_set_channel_callbacks(self._libssh_channel, &cb)
 
-        libssh.ssh_channel_send_eof(channel)
-        result.returncode = libssh.ssh_channel_get_exit_status(channel)
+        libssh.ssh_channel_send_eof(self._libssh_channel)
 
-        libssh.ssh_channel_free(channel)
+        result.returncode = self.get_channel_exit_status()
 
         return result
 
     def get_channel_exit_status(self):
         return libssh.ssh_channel_get_exit_status(self._libssh_channel)
 
     def close(self):
-        libssh.ssh_channel_free(self._libssh_channel)
-        self._libssh_channel = NULL
+        if self._libssh_channel is not NULL:
+            libssh.ssh_channel_close(self._libssh_channel)
```

Comment:
This doesn't look necessary. The API states that `ssh_channel_free()` does `ssh_channel_close()` internally.

Reply:
https://api.libssh.org/stable/libssh_tutor_command.html the libssh examples uses free and close API's separately hence added it to be on safe side

Reply:
I think they've forgotten to update the examples. I checked the source:
https://git.libssh.org/projects/libssh.git/tree/src/channels.c#n1153. "free" actually has clever logic deciding whether to call "close".

Reply:
OTOH maybe it's better to call it unconditionally...